mongo/options/autoencryptionoptions.go

```diff
@@ -91,6 +96,38 @@ func (a *AutoEncryptionOptions) SetExtraOptions(extraOpts map[string]interface{}
 	return a
 }
 
+// SetTLSOptions specifies tls.Config instances for each KMS provider to use to configure TLS on all connections created
+// to the cluster. The input map should contain a mapping from each KMS provider to a document containing the necessary 
+// options, as follows:
+//
+// {
+//		"kmip": {
+//			"tlsCertificateKeyFile": "foo.pem",
+// 			"tlsCAFile": "fooCA.pem"
+//		}
+// }
+//
+// Currently, the following TLS options are supported:
+//
+// 1. "tlsCertificateKeyFile" (or "sslClientCertificateKeyFile"): The "tlsCertificateKeyFile" option specifies a path to 
+// the client certificate and private key, which must be concatenated into one file. 
+//
+// 2. "tlsCertificateKeyFilePassword" (or "sslClientCertificateKeyPassword"): Specify the password to decrypt the client
+// private key file (e.g. "tlsCertificateKeyFilePassword=password").
+//
+// 3. "tlsCaFile" (or "sslCertificateAuthorityFile"): Specify the path to a single or bundle of certificate authorities
+// to be considered trusted when making a TLS connection (e.g. "tlsCaFile=/path/to/caFile").
+//
+// This should only be used to set custom TLS options. By default, the connection will use an empty tls.Config{}.
+func (a *AutoEncryptionOptions) SetTLSOptions(tlsOpts map[string]map[string]interface{}) (*AutoEncryptionOptions, error) {
```

Comment:
Also wanted to call out that `SetTLSOptions()` currently returns both an options object and an error, unlike other options mutators in the same file. This complicates chaining -- would it better to only return the options object and print the error/panic?

Reply:
I agree that having a `Set`ter return an error is not particularly idiomatic (as you mentioned it stops users from chaining). `panic`ing, from what I've gathered, is almost always less desirable than returning an error to the user, though. And, we pretty much never use `fmt` to print errors to stderr or stdout to "warn" users. This function signature might just be an unfortunate side effect of `tlsOpts` being `map[string]map[string]interface{}` instead of a concrete type.

Reply:
Here's how the current API would be used:
```go
var tlsOpts map[string]map[string]interface{}
aeOpts, err := options.AutoEncryption().SetTLSOptions(tlsOpts)
if err != nil {
	panic(err)
}
opts := options.Client().SetAutoEncryptionOptions(aeOpts)
client, err := mongo.Connect(context.TODO(), opts)
if err != nil {
	panic(err)
}
```
It's usable, but mostly requires that `SetTLSOptions` is the last call in the chain to make sense. If we ever added another option function that returned an error, the usability pattern would break.

Here's an example of a possible API replacing `SetTLSOptions` with `SetTLSConfig` and using a separate `BuildTLSConfig` that works like `SetTLSOptions` but just returns the `tls.Config`:
```go
var tlsOpts map[string]map[string]interface{}
cfg, err := options.BuildTLSConfig(tlsOpts)
if err != nil {
	panic(err)
}
aeOpts := options.AutoEncryption().SetTLSConfig(cfg)
opts := options.Client().SetAutoEncryptionOptions(aeOpts)
client, err := mongo.Connect(context.TODO(), opts)
if err != nil {
	panic(err)
}
```

Here's an example of a possible API that stores errors on the `AutoEncryptionOptions`/`ClientEncryptionOptions` structs and handles them as part of `Client` or `ClientEncryption` creation:
```go
var tlsOpts map[string]map[string]interface{}
aeOpts := options.AutoEncryption().SetTLSOptions(tlsOpts)
opts := options.Client().SetAutoEncryptionOptions(aeOpts)
client, err := mongo.Connect(context.TODO(), opts)
if err != nil {
	panic(err)
}
```
Note that to implement that API, you'd have to add a `Validate` function to `AutoEncryptionOptions`/`ClientEncryptionOptions` and update everywhere that accepts an `AutoEncryptionOptions`/`ClientEncryptionOptions` to call `Validate` and handle errors.

The current API or `BuildTLSConfig`+`SetTLSConfig` are similarly usable. `BuildTLSConfig`+`SetTLSConfig` has the advantage that it follows the current API pattern and is simple to implement (`BuildTLSConfig` is basically `applyTLSOptions`). Storing errors on the structs and handling them later would be significantly more difficult to implement, so it might not be worth it just to get rid of one error check.

Optional: Consider using the `BuildTLSConfig`+`SetTLSConfig` pattern.